sdk/samples/core/mosquitto_multiple_rpc_sample.c

```diff
@@ -400,51 +436,36 @@ az_result mqtt_callback(az_mqtt5_connection* client, az_event event, void* callb
   return AZ_OK;
 }
 
-/**
- * @brief Removes any expired commands from the client_pending_commands array
- * @note Even if a command has expired, if we get a response for it, we will still receive an event
- * with the results in the mqtt_callback
- */
-void remove_expired_commands()
-{
-  pending_command* expired_command = get_first_expired_command(client_pending_commands);
-  while (expired_command != NULL)
-  {
-    printf(LOG_APP_ERROR "command ");
-    print_correlation_id(expired_command->correlation_id);
-    printf(" expired\n");
-    az_result ret = remove_command(&client_pending_commands, expired_command->correlation_id);
-    if (ret != AZ_OK)
-    {
-      printf(LOG_APP_ERROR "Expired command not a pending command\n");
-    }
-    expired_command = get_first_expired_command(client_pending_commands);
-  }
-}
-
 az_result invoke_begin(az_span command_name, az_span payload)
 {
-  uuid_t new_uuid;
-  uuid_generate(new_uuid);
+  // Remove any faulted commands before adding a new one to avoid exceeding the max pending requests
+  // unnecessarily
+  remove_faulted_commands();
+  // this memory must be free'd once the application is done with the command with
+  // remove_and_free_command()
+  uuid_t* new_uuid = malloc(AZ_MQTT5_RPC_CORRELATION_ID_LENGTH + 1);
+  az_mqtt5_request* request = malloc(sizeof(az_mqtt5_request));
+
+  uuid_generate(*new_uuid);
+
   az_mqtt5_rpc_client_invoke_req_event_data command_data
-      = { .correlation_id = az_span_create((uint8_t*)new_uuid, AZ_MQTT5_RPC_CORRELATION_ID_LENGTH),
+      = { .correlation_id = az_span_create((uint8_t*)*new_uuid, AZ_MQTT5_RPC_CORRELATION_ID_LENGTH),
           .content_type = content_type,
+          .request_memory = request,
           .rpc_server_client_id = server_client_id,
           .command_name = command_name,
-          .request_payload = payload };
-  LOG_AND_EXIT_IF_FAILED(add_command(
-      &client_pending_commands,
-      command_data.correlation_id,
-      command_name,
-      CLIENT_COMMAND_TIMEOUT_MS));
+          .request_payload = payload,
+          .timeout_s = CLIENT_COMMAND_TIMEOUT_S };
   az_result rc = az_mqtt5_rpc_client_invoke_begin(&rpc_client, &command_data);
```

Comment:
no, because the application must free the memory associated with the faulted command. I do remove faulted commands at the beginning of this function (invoke_begin) though!